crates/ruff_server/src/session/index.rs

```diff
@@ -0,0 +1,441 @@
+use anyhow::anyhow;
+use rustc_hash::FxHashMap;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use crate::{
+    edit::{DocumentKey, DocumentVersion, NotebookDocument},
+    PositionEncoding, TextDocument,
+};
+
+use super::{
+    settings::{self, ResolvedClientSettings},
+    ClientSettings,
+};
+
+mod ruff_settings;
+
+pub(crate) use ruff_settings::RuffSettings;
+
+type DocumentIndex = FxHashMap<PathBuf, DocumentController>;
+type NotebookCellIndex = FxHashMap<lsp_types::Url, PathBuf>;
+type SettingsIndex = BTreeMap<PathBuf, WorkspaceSettings>;
+
+/// Stores and tracks all open documents in a session, along with their associated settings.
+#[derive(Default)]
+pub(crate) struct Index {
+    /// Maps all document file paths to the associated document controller
+    documents: DocumentIndex,
+    /// Maps opaque cell URLs to a notebook path
+    notebook_cells: NotebookCellIndex,
+    /// Maps a workspace folder root to its settings.
+    settings: SettingsIndex,
+}
+
+/// Settings associated with a workspace.
+struct WorkspaceSettings {
+    client_settings: ResolvedClientSettings,
+    workspace_settings_index: ruff_settings::RuffSettingsIndex,
+}
+
+/// A mutable handler to an underlying document.
+enum DocumentController {
+    Text(Arc<TextDocument>),
+    Notebook(Arc<NotebookDocument>),
+}
+
+/// A read-only query to an open document.
+/// This query can 'select' a text document, full notebook, or notebook cell.
+/// It also includes document settings.
+#[derive(Clone)]
+pub(crate) enum DocumentQuery {
+    Text {
+        file_path: PathBuf,
+        document: Arc<TextDocument>,
+        settings: Arc<RuffSettings>,
+    },
+    Notebook {
+        /// The selected notebook cell.
+        /// If this is `None`, the entire notebook file is selected.
+        cell_uri: Option<lsp_types::Url>,
+        file_path: PathBuf,
+        notebook: Arc<NotebookDocument>,
+        settings: Arc<RuffSettings>,
+    },
```

Comment:
By the comment, do you mean that the `Notebook` variant either represents a specific cell or an entire Notebook? I'd expand your comment, especially what does "entire notebook file is selected" mean.

Reply:
> By the comment, do you mean that the Notebook variant either represents a specific cell or an entire Notebook?

That's correct - I'll try to clarify this.